src/protocols/secure_channel/CheckinMessage.cpp

```diff
@@ -0,0 +1,110 @@
+    Encoding::LittleEndian::Write32(appDataStartPtr, counter);
+    appDataStartPtr -= sizeof(uint32_t);
+
+    {
+        chip::Crypto::HMAC_sha shaHandler;
+        uint8_t nonceWorkBuffer[CHIP_CRYPTO_HASH_LEN_BYTES] = { 0 };
+        ReturnErrorOnFailure(shaHandler.HMAC_SHA256(key.As<Aes128KeyByteArray>(), sizeof(Aes128KeyByteArray), appDataStartPtr,
+                                                    sizeof(counter), nonceWorkBuffer, CHIP_CRYPTO_HASH_LEN_BYTES));
+        memcpy(output.data(), nonceWorkBuffer, CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES);
+    }
+
+    // In place encryption to save some RAM
+    memcpy(appDataStartPtr + sizeof(counter), appData.data(), appData.size());
+
+    // API constraint. Mic buffer needs to be separated.
+    uint8_t mic[CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES];
+    uint8_t * micPtr = appDataStartPtr + appData.size() + sizeof(counter);
+
+    ReturnErrorOnFailure(Crypto::AES_CCM_encrypt(appDataStartPtr, appData.size() + sizeof(counter), nullptr, 0, key, output.data(),
+                                                 CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES, appDataStartPtr, mic,
+                                                 CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES));
+
+    memcpy(micPtr, mic, CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES);
+
+    if (payloadSize != nullptr)
+    {
+        *payloadSize = static_cast<uint16_t>(appData.size() + sMinPayloadSize);
+    }
+
+    return err;
+}
+
+CHIP_ERROR CheckinMessage::ParseCheckinMessagePayload(Crypto::Aes128KeyHandle & key, ByteSpan & payload, uint32_t & counter,
+                                                      MutableByteSpan & appData)
+{
+    VerifyOrReturnError(payload.size() >= sMinPayloadSize, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(payload.size() <= (sMinPayloadSize + CHIP_CHECK_IN_APP_DATA_MAX_SIZE), CHIP_ERROR_INVALID_ARGUMENT);
+
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+    uint16_t appDataSize = GetAppDataSize(payload);
+
+    // To prevent workbuffer usage, appData size need to be large enough to hold both the appData and the counter
+    VerifyOrReturnError(appData.size() >= appDataSize + sizeof(uint32_t), CHIP_ERROR_INVALID_ARGUMENT);
+
+    ByteSpan nonce         = payload.SubSpan(0, CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES);
+    ByteSpan encryptedData = payload.SubSpan(CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES, appDataSize + sizeof(counter));
+    ByteSpan mic =
+        payload.SubSpan(CHIP_CRYPTO_AEAD_NONCE_LENGTH_BYTES + appDataSize + sizeof(counter), CHIP_CRYPTO_AEAD_MIC_LENGTH_BYTES);
+
+    err = Crypto::AES_CCM_decrypt(encryptedData.data(), encryptedData.size(), nullptr, 0, mic.data(), mic.size(), key, nonce.data(),
+                                  nonce.size(), appData.data());
+
+    ReturnErrorOnFailure(err);
+
+    counter = Encoding::LittleEndian::Get32(appData.data());
+    // Shift to remove the counter from the appData
+    memcpy(appData.data(), appData.data() + sizeof(uint32_t), appDataSize);
```

Comment:
Where are we updating appData.size()?  We should be doing that, no?

I don't understand how this is passing tests....

Reply:
added resized